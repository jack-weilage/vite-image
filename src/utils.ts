import type { ImageConfig, Transformer, PluginConfig } from '../types'
import type { Sharp, Metadata } from 'sharp'

import { type BinaryLike, createHash } from 'crypto'
import { basename, extname } from 'path'
import { CONFIG_SCHEMA } from './constants'

/** Limit a number to between min and max. */
export const minmax = (num: number, min: number, max: number) => Math.max(Math.min(num, max), min)
/** 
 * Create a SHA1 hash from a string.
 * 
 * SHA256 is more secure, but speed matters more here.
*/
export const create_hash = (str: BinaryLike): string => createHash('sha1').update(str).digest('hex')
/** Extract the name of a file, without its extension. */
export const filename = (path: string) => basename(path, extname(path))
/** De-duplicate an array. */
export const dedupe = <T>(arr: T[]) => [ ...new Set(arr) ]

/** Create a `Partial` of any object. */
export function copy_only_keys<T>(obj: T, keys: (keyof T)[]): Partial<T> {
    const partial = {} as Partial<T>
    for (const key of keys)
        partial[key] = obj[key]

    return partial
}

/** Validate and combine a user config and default config. */
export function parse_config(user_config: Partial<PluginConfig>, default_config: PluginConfig): PluginConfig
{
    const config = Object.assign({ ...default_config }, user_config)

    const errors = CONFIG_SCHEMA.validate(config)
    if (errors.length !== 0)
        throw new AggregateError(errors)
    
    return config
}

/** Coerces values to string | number | boolean. */
const format_value = (val: string) => {
    if (val === '' || val === 'true')
        return true

    if (val === 'false')
        return false
    
    if (!Number.isNaN(+val))
        return +val

    return val
}

//TODO: Handle `!blur` as `blur=false`
/** Create unique configs from arrays of values. */
export function create_configs(params: URLSearchParams, deliminator: string): ImageConfig[]
{
    const aggregated = {} as Record<string, (string | number | boolean)[]>
    for (const key of dedupe([ ...params.keys() ]))
    {
        // Get every occurrence of the key, then split by the deliminator.
        const values = params.getAll(key)
            .flatMap(value => value.split(deliminator))
            .map(format_value)

        aggregated[key] = dedupe(values)
    }
    
    const keys = Object.keys(aggregated)
    const values = Object.values(aggregated)

    // If there is only one key, the code below won't work.
    if (keys.length === 1)
        return values[0].map(value => ({ [keys[0]]: value }))
    
    // Create an array of unique configs.
    // Each value will have the index of the corresponding keys.
    const groups = values
        //@ts-expect-error: This reducer will take in (string | number | boolean)[] and return (string | number | boolean)[][]
        .reduce((acc, cur) => acc.flatMap(a => cur.map(b => [ a, b ].flat()))) as any as (string | number | boolean)[][]
    
    const configs = [] as ImageConfig[]
    for (const group of groups) {
        const config = {} as Record<string, string | number | boolean>

        // For every value, assign the corresponding key.
        for (let i = 0; i < keys.length; i++)
            config[keys[i]] = group[i]
        
        configs.push(config)
    }

    return configs
}

/** Apply all transformers to an image. */
export function apply_transformers(image: Sharp, metadata: Metadata, config: ImageConfig, transformers: Transformer[])
{
    const applied_transformers = []

    for (const { matcher, transform, name } of transformers)
    {
        if (!matcher(config))
            continue
    
        applied_transformers.push(name)
        image = transform(image, config, metadata)
    }

    return { image, applied_transformers }
}