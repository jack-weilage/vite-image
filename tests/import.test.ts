import { Window } from 'happy-dom'
import { test } from './utils'
import { it, expect, describe, beforeEach } from 'vitest'


let window: Window
beforeEach(() => {
    window = new Window()
})

describe('imports', () => {
    it.each([
        // Just a normal import.
        [ './images/dog.jpg', '/assets/dog.1b15ce03.jpg' ],
        // A little fancy. An import with searchparams
        [ './images/dog.jpg?foo=bar', '/assets/dog.1b15ce03.jpg?foo=bar' ],
        // Even fancier. An import with searchparams and hash
        [ './images/dog.jpg?foo=bar#baz-qux', '/assets/dog.1b15ce03.jpg?foo=bar#baz-qux' ]
    ])('does not affect default imports', async (input, output) => expect(await test(window, input)).toBe(output))
    it.each([
        [ './images/dog.jpg?width=600,1200&height=800,1600', [
            [ 600, 800 ],
            [ 600, 1600 ],
            [ 1200, 800 ],
            [ 1200, 1600 ]
        ] ],
        [ './images/dog.jpg?width=20,400,30&height=50', [
            [ 20, 50 ],
            [ 400, 50 ],
            [ 30, 50 ]
        ] ],
        [ './images/dog.jpg?width=500,800,1000', [
            [ 500, 375 ],
            [ 800, 600 ],
            [ 1000, 750 ]
        ] ],
        [ './images/dog.jpg?height=200,900,1000', [
            [ 267, 200 ],
            [ 1200, 900 ],
            [ 1333, 1000 ]
        ] ]
    ])('affects custom imports', async (input, output) => {
        const data = test(window, input)
            .then(images => images.map(img => [ img.width, img.height ]))
        
        expect(await data).toEqual(output)
    })
})