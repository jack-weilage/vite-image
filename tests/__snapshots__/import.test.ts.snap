// Vitest Snapshot v1

exports[`affects custom imports: !flip 1`] = `
[
  "/assets/dog-9a9fbf08.jpeg",
]
`;

exports[`affects custom imports: blur 1`] = `
[
  "/assets/dog-633250ad.jpeg",
]
`;

exports[`affects custom imports: blur&!flip 1`] = `
[
  "/assets/dog-633250ad.jpeg",
]
`;

exports[`affects custom imports: blur=true 1`] = `
[
  "/assets/dog-633250ad.jpeg",
]
`;

exports[`affects custom imports: width=500,600&height=800&width=400 1`] = `
[
  "/assets/dog-f9bb58b3.jpeg",
  "/assets/dog-88a85387.jpeg",
  "/assets/dog-adabbc87.jpeg",
]
`;

exports[`affects custom imports: width=600 1`] = `
[
  "/assets/dog-88a85387.jpeg",
]
`;

exports[`affects custom imports: width=600&blur&height=400 1`] = `
[
  "/assets/dog-6ca67b9c.jpeg",
]
`;

exports[`affects custom imports: width=600&height=400 1`] = `
[
  "/assets/dog-2d621bfe.jpeg",
]
`;

exports[`affects custom imports: width=600,800&height=400,500 1`] = `
[
  "/assets/dog-2d621bfe.jpeg",
  "/assets/dog-af2d5f89.jpeg",
  "/assets/dog-81677fe4.jpeg",
  "/assets/dog-af0e2b48.jpeg",
]
`;

exports[`affects custom imports: width=600,800,1000 1`] = `
[
  "/assets/dog-88a85387.jpeg",
  "/assets/dog-98297939.jpeg",
  "/assets/dog-c851a66b.jpeg",
]
`;

exports[`does not affect default imports:  1`] = `"/assets/dog-1b15ce03.jpg"`;

exports[`does not affect default imports: ?foo=bar 1`] = `"/assets/dog-1b15ce03.jpg?foo=bar"`;

exports[`does not affect default imports: ?foo=bar#baz-qux 1`] = `"/assets/dog-1b15ce03.jpg?foo=bar#baz-qux"`;
